static/app/views/dashboardsV2/layoutUtils.tsx

```diff
@@ -75,3 +87,74 @@ export function getDashboardLayout(widgets: Widget[]): Layout[] {
       i: constructGridItemKey(widget),
     }));
 }
+
+export function pickDefinedStoreKeys(layout: Layout): Partial<Layout> {
+  return pickBy(layout, (value, key) => defined(value) && STORE_KEYS.includes(key));
+}
+
+export function getWidgetHeight(displayType: DisplayType): number {
+  return displayType === DisplayType.BIG_NUMBER ? 1 : 2;
+}
+
+/**
+ * Creates an array from layouts where each column stores how deep it is.
+ */
+export function calculateColumnDepths(layouts: Layout[]): number[] {
+  const depths = Array(NUM_DESKTOP_COLS).fill(0);
+
+  // For each layout's x, record the max depth
+  layouts
+    .filter(({i}) => i !== ADD_WIDGET_BUTTON_DRAG_ID)
+    .forEach(({x, w, y, h}) => {
+      // Adjust the column depths for each column the widget takes up
+      for (let col = x; col < x + w; col++) {
+        depths[col] = Math.max(y + h, depths[col]);
+      }
+    });
+
+  return depths;
+}
+
+/**
+ * Find the next place to place a widget and also returns the next
+ * input when this operation needs to be called multiple times.
+ *
+ * @param columnDepths A profile of how deep the widgets in a column extend.
+ * @param height The desired height of the next widget we want to place.
+ * @returns An {x, y} positioning for the next available spot, as well as the
+ * next columnDepths array if this position were used.
+ */
+export function getNextAvailablePosition(
+  columnDepths: number[],
+  height: number
+): NextPosition {
+  const maxColumnDepth = Math.max(...columnDepths);
+
+  // Look for an opening at each depth by scanning from 0, 0
+  // By scanning from 0 depth to the highest depth, we ensure
+  // we get the top-most available spot
+  for (let currDepth = 0; currDepth <= maxColumnDepth; currDepth++) {
+    for (let start = 0; start <= columnDepths.length - DEFAULT_WIDGET_WIDTH; start++) {
+      if (columnDepths[start] > currDepth) {
+        // There are potentially widgets in the way here, so skip
+        continue;
+      }
+
+      // If all of the columns from start to end (the size of the widget)
+      // have at most the current depth, then we've found a valid positioning
+      // No other widgets extend into the space we need
+      const end = start + DEFAULT_WIDGET_WIDTH;
+      if (columnDepths.slice(start, end).every(val => val <= currDepth)) {
+        for (let col = start; col < start + DEFAULT_WIDGET_WIDTH; col++) {
+          columnDepths[col] = currDepth + height;
```

Comment:
I think this is going to mutate the array passed in the input. This probably won't actually cause any issues but it could be a little cleaner to clone `columnDepths` at the start of the function and work off of that instead to preserve idempotence.

Reply:
Fixed! Thanks for calling that out. I renamed the param to `initialColumnDepths` and copy that as `columnDepths` for the rest of the work